### PR TITLE
Accept numericoid hexstring x509

### LINKED ISSUE #7924: RFC 4514: Escape special characters
Subtask of #6785

Implement escaping and de-escaping of special characters as described in [section 2.4 of RFC 4514](https://www.rfc-editor.org/rfc/rfc4514#section-2.4).

-  Modify `mbedtls_x509_dn_gets()` to escape special characters where it does not do so already.
- Modify `mbedtls_x509_string_to_names()` to de-escape special characters.
- Add testing for both functions to cover cases of escaped characters and (for parsing) incorrectly escaped characters.

The escaping of hex pairs (e.g. `\A5`, `\FF`) is covered in a separate issue.

### LINKED ISSUE #7926: RFC 4514: Escape hex pairs
Subtask of #6785

Follows #7924

Implement escaping and de-escaping of hex pairs as specified in RFC 4514 sections [2.4](https://www.rfc-editor.org/rfc/rfc4514#section-2.4) and [3](https://www.rfc-editor.org/rfc/rfc4514#section-3).

-  Modify `mbedtls_x509_dn_gets()` to escape any non-ascii characters as hex pairs e.g. `\AB\CD`.
- Modify `mbedtls_x509_string_to_names()` to accept escaped hex-pairs.
- Add testing for this functionality in both functions.

### LINKED ISSUE #7928: RFC 4514: Support numericoid / hexstring representation
Subtask of #6785.

Follows #7927

Add support for the representation of an AttributeType/AttributeValue pair in the numericoid and hexstring representation as described in [RFC 4514 section 3](https://www.rfc-editor.org/rfc/rfc4514#section-3).

For example, an AttributeType/AttributeValue pair could be represented as: `1.3.6.1.4.1.1466.0=#04024869`.

- Modify `mbedtls_x509_string_to_names()` to accept this representation, add tests.
- Modify `mbedtls_x509_dn_gets()` to use this representation when the `tag` of the value in the `mbedtls_x509_name` is not a string type, add tests.

### PR DESCRIPTION
**Pull request by AgathiyanB**
## Description

fix #7924  
fix #7926  
fix #7928 

Working towards #6785 

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** in a future PR
- [x] **backport** not required as new feature
- [x] **tests** added

### CODE REVIEW

File: tests/suites/test_suite_x509write.data [hunk 1]
```diff
@@ -184,8 +184,38 @@ mbedtls_x509_string_to_names:"C=NL, O=Offspark\\a Inc., OU=PolarSSL":"":MBEDTLS_
...
+X509 String to Names #11 (Escape ascii hexpairs uppercase encoded)
+mbedtls_x509_string_to_names:"C=NL, O=\\4F\\66\\66\\73\\70\\61\\72\\6B, OU=PolarSSL":"C=NL, O=Offspark, OU=PolarSSL":0
+
+X509 String to Names #12 (Escape ascii hexpairs lowercase encoded)
+mbedtls_x509_string_to_names:"C=NL, O=\\4f\\66\\66\\73\\70\\61\\72\\6b, OU=PolarSSL":"C=NL, O=Offspark, OU=PolarSSL":0
+
+X509 String to Names #13 (Invalid hexpair escape at end of string)
+mbedtls_x509_string_to_names:"C=NL, O=\\4f\\66\\66\\73\\70\\61\\72\\6, OU=PolarSSL":"":MBEDTLS_ERR_X509_INVALID_NAME
+
+X509 String to Names #14 (Escape non-ascii hexpairs)
```
Comment (davidhorstmann-arm) [inline tests/suites/test_suite_x509write.data:208]: Above I've noted that it would be best to reject strings with escaped `'\0'` characters. Please change this test to a negative test and add a positive test for an escaped character that is non-zero

File: tests/suites/test_suite_x509write.data [hunk 2]
```diff
@@ -184,8 +184,38 @@ mbedtls_x509_string_to_names:"C=NL, O=Offspark\\a Inc., OU=PolarSSL":"":MBEDTLS_
...
+
+X509 String to Names #15 (Invalid hexpairs)
+mbedtls_x509_string_to_names:"C=NL, O=Of\\flspark, OU=PolarSSL":"":MBEDTLS_ERR_X509_INVALID_NAME
+
+X509 String to Names #16 (Accept numercoid/hexstring)
+mbedtls_x509_string_to_names:"C=NL, 2.5.4.10=#0C084F6666737061726B, OU=PolarSSL":"C=NL, O=Offspark, OU=PolarSSL":0
+
+X509 String to Names #17 (Accept numercoid/hexstring, output as bitstring)
+mbedtls_x509_string_to_names:"C=NL, 2.5.4.10=#03084F6666737061726B, OU=PolarSSL":"C=NL, O=#03084F6666737061726B, OU=PolarSSL":0
+
```
Comment (davidhorstmann-arm) [inline tests/suites/test_suite_x509write.data:219]: It would be good to have some more negative testing of the numericoid / hexstring format, e.g.:
* OIDs that are corrupted / invalid
* Hexstrings with an odd number of characters
* Hexstrings that are not valid ASN.1

File: library/x509_create.c [hunk 3]
```diff
@@ -123,53 +130,199 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
...
+    c++;
+    if ((*tag = hexpair_to_int(*c, *(c+1))) == -1) {
+        return MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+    c += 2;
+    p = asn1_der_buf;
+    for (p = asn1_der_buf; c < end; c += 2) {
+        if ((c + 1 >= end) || (n = hexpair_to_int(*c, *(c+1))) == -1) {
+            return MBEDTLS_ERR_X509_INVALID_NAME;
+        }
```
Comment (davidhorstmann-arm) [inline library/x509_create.c:245]: For now I think we should explicitly reject escaped `'\0'` characters. Allowing the user to create certificates with strings that contain `'\0'` in the middle seems like a bad idea. UTF-8 work can address unacceptable characters more thoroughly, but it's best to have some safeguard.
```suggestion
        }
        if (n == '\0') {
            return MBEDTLS_ERR_X509_INVALID_NAME;
        }
```

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
Quite a few comments, but the overall approach looks good!

File: library/x509_create.c (hunk 3 above)
Comment (AgathiyanB) [reply library/x509_create.c:245]: This function doesn't just parse for strings but any ber encoded data, I know some certificates in the testing framework have encoded bitstrings with null bytes.

File: library/x509_create.c [hunk 4]
```diff
@@ -123,53 +130,199 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
...
+    unsigned char *d = data;
+    int n;
+    while (c < end) {
+        if (*c == '\\') {
+            c++;
+
+            /* Check for valid escaped characters in RFC 4514 in Section 3*/
+            if (c + 1 < end && (n = hexpair_to_int(*c, *(c+1))) != -1) {
+                hexpair = 1;
+                *(d++) = n;
```
Comment (davidhorstmann-arm) [inline library/x509_create.c:198]: For now I think we should explicitly reject escaped `'\0'` characters. Allowing the user to create certificates with strings that contain `'\0'` in the middle seems like a bad idea. UTF-8 work can address unacceptable characters more thoroughly, but it's best to have some safeguard.
```suggestion
                if (n == '\0') {
                    return MBEDTLS_ERR_X509_INVALID_NAME;
                }
                *(d++) = n;
```

File: library/x509_create.c (hunk 3 above)
Comment (davidhorstmann-arm) [reply library/x509_create.c:245]: You're right, I meant to raise this [above](https://github.com/Mbed-TLS/mbedtls/pull/8025/files/d265d14c4b56959333d0e36e0b257a788fb22313#r1290375707), not here.

File: tests/suites/test_suite_x509write.data (hunk 2 above)
Comment (AgathiyanB) [reply tests/suites/test_suite_x509write.data:219]: By non valid ASN.1, should we check the length matches up

**Comment by tom-cosgrove-arm** [comment]
@AgathiyanB I see you've added more commits, but the `needs-work` label is still present. When this is ready for review, can you remove `needs-work`?

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
It's looking in good shape apart from a few things

File: tests/suites/test_suite_x509write.data [hunk 5]
```diff
@@ -214,7 +214,7 @@ mbedtls_x509_string_to_names:"C=NL, O=Of\\flspark, OU=PolarSSL":"":MBEDTLS_ERR_X
 X509 String to Names #16 (Accept numercoid/hexstring)
 mbedtls_x509_string_to_names:"C=NL, 2.5.4.10=#0C084F6666737061726B, OU=PolarSSL":"C=NL, O=Offspark, OU=PolarSSL":0
 
-X509 String to Names #17 (Accept numercoid/hexstring, output as bitstring)
+X509 String to Names #17 (Output attributetype as bitstring)
```
[the remark this answers is not in the document]
Comment (AgathiyanB) [reply tests/suites/test_suite_x509write.data]: This test no longer exists as it was more suitable in x509parse

[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply tests/suites/test_suite_x509write.data]: Ah excellent, no problem then!

File: library/x509_create.c [hunk 6]
```diff
@@ -273,10 +273,18 @@ int mbedtls_x509_string_to_names(mbedtls_asn1_named_data **head, const char *nam
         }
 
         if (!in_attr_type && ((*c == ',' && *(c-1) != '\\') || c == end)) {
+            if (*s == '#') {
```
Comment (davidhorstmann-arm) [inline library/x509_create.c:276]: Thinking about it, we need to check that there actually is something in the attribute value, otherwise we can pass `CN=,` and cause a 1-byte buffer overread.
```suggestion
            if (s == c) {
                /* Empty AttributeValue */
                mbedtls_free(oid.p);
                return MBEDTLS_ERR_X509_INVALID_NAME;
            } else if (*s == '#') {
```

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
Looks good apart from one last thing

File: library/x509_create.c (hunk 6 above)
Comment (tom-cosgrove-arm) [reply library/x509_create.c:276]: Note: there should be a test for this case

File: tests/suites/test_suite_x509write.data [hunk 7]
```diff
@@ -184,8 +184,47 @@ mbedtls_x509_string_to_names:"C=NL, O=Offspark\\a Inc., OU=PolarSSL":"":MBEDTLS_
...
+mbedtls_x509_string_to_names:"C=NL, 10.5.4.10=#0C084F6666737061726B, OU=PolarSSL":"":MBEDTLS_ERR_X509_INVALID_NAME
+
+X509 String to Names #18 (short name and hexstring)
+mbedtls_x509_string_to_names:"C=NL, O=#0C084F6666737061726B, OU=PolarSSL":"C=NL, O=Offspark, OU=PolarSSL":0
+
+X509 String to Names #19 (Accept non-ascii hexpairs)
+mbedtls_x509_string_to_names:"C=NL, O=Of\\CCspark, OU=PolarSSL":"C=NL, O=Of\\CCspark, OU=PolarSSL":0
+
+X509 String to Names #20 (Reject empty AttributeValue)
+mbedtls_x509_string_to_names:"C=NL, O=, OU=PolarSSL":"":MBEDTLS_ERR_X509_INVALID_NAME
```
Comment (davidhorstmann-arm) [inline tests/suites/test_suite_x509write.data:227]: Trailing whitespace here causing `check_files.py` to fail

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
LGTM apart from 2 things:

1. Trailing whitespace is causing a CI failure (see comment)
2. I've realised that this PR needs a ChangeLog entry of its own, since this is going to get into the release but the next PR won't. Apologies for not spotting this earlier.

**Review by davidhorstmann-arm — DISMISSED** [review]
LGTM, thanks!

File: include/mbedtls/asn1.h [hunk 8]
```diff
@@ -97,14 +97,13 @@
...
 #define MBEDTLS_ASN1_IS_STRING_TAG(tag)                                     \
-    ((tag) < 32u && (                                                      \
-         ((1u << (tag)) & ((1u << MBEDTLS_ASN1_BMP_STRING)       |     \
-                           (1u << MBEDTLS_ASN1_UTF8_STRING)      |     \
-                           (1u << MBEDTLS_ASN1_T61_STRING)       |     \
-                           (1u << MBEDTLS_ASN1_IA5_STRING)       |     \
-                           (1u << MBEDTLS_ASN1_UNIVERSAL_STRING) |     \
-                           (1u << MBEDTLS_ASN1_PRINTABLE_STRING) |     \
-                           (1u << MBEDTLS_ASN1_BIT_STRING))) != 0))
+    ((tag) < 32 && (                                                      \
```
Comment (tom-cosgrove-arm) [inline include/mbedtls/asn1.h:100]: I really don't like this change, since if `tag` has `-1` this test will now pass and we'll attempt to shift left by `-1`.

Instead, I think a better fix might be to cast `tag` to unsigned for the comparison against `32u`.

Also, the changes from `1u` to `1` aren't necessary, and could be dangerous: `(signed) 1 << 31` is UB, IIRC

Comment (tom-cosgrove-arm) [reply include/mbedtls/asn1.h:100]: and (minor): if we're modifying this, let's have all the `\`s line up :)

File: library/x509_create.c [hunk 9]
```diff
@@ -123,79 +130,206 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
...
+static int parse_attribute_value_string(const char *s,
+                                        int len,
+                                        unsigned char *data,
+                                        size_t *data_len)
+{
+    const char *c = s;
+    const char *end = c + len;
+    int hexpair = 0;
+    unsigned char *d = data;
+    int n;
```
Comment (tom-cosgrove-arm) [inline library/x509_create.c:160]: blank line between variable declarations and code

File: library/x509_create.c [hunk 10]
```diff
@@ -123,79 +130,206 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
...
+    while (c < end) {
+        if (*c == '\\') {
+            c++;
+
+            /* Check for valid escaped characters in RFC 4514 in Section 3*/
```
Comment (tom-cosgrove-arm) [inline library/x509_create.c:165]: (minor) probably should be something like (also note space before `*/`)

```suggestion
            /* Check for valid escaped characters as per RFC 4514 Section 3 */
```

File: library/x509_create.c [hunk 11]
```diff
@@ -123,79 +130,206 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
...
+                                        int len,
+                                        unsigned char *data,
+                                        size_t *data_len)
+{
+    const char *c = s;
+    const char *end = c + len;
+    int hexpair = 0;
+    unsigned char *d = data;
+    int n;
+    while (c < end) {
```
Comment (tom-cosgrove-arm) [inline library/x509_create.c:161]: It feels like this might be better off as a `for`-loop, and then the use of `hexpair = 1` could be replaced with a `continue` - have you tried that?

File: library/x509_create.c [hunk 12]
```diff
@@ -123,79 +130,206 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
...
 }
 
+static int hex_to_int(char c)
+{
+    return ('0' <= c && c <= '9') ? (c - '0') :
+           ('a' <= c && c <= 'f') ? (c - 'a' + 10) :
+           ('A' <= c && c <= 'F') ? (c - 'A' + 10) : -1;
+}
+
+static int hexpair_to_int(char c1, char c2)
```
Comment (tom-cosgrove-arm) [inline library/x509_create.c:140]: `hexpair_to_int()` is only ever called with `*p` and `*(p + 1)`, so would it be better to pass just `p` to it?

File: library/x509_create.c [hunk 13]
```diff
@@ -123,79 +130,206 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
...
+{
+    int n1 = hex_to_int(c1);
+    int n2 = hex_to_int(c2);
```
Comment (tom-cosgrove-arm) [inline library/x509_create.c:143]: This function would look nicer with a blank line after this line

**Comment by tom-cosgrove-arm** [comment]
In addition to those comments, I think it would be good to have some "round-trip" tests, especially with characters that get escaped

File: library/x509.c [hunk 14]
```diff
@@ -840,32 +853,95 @@ int mbedtls_x509_dn_gets(char *buf, size_t size, const mbedtls_x509_name *dn)
...
+                    return MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
+                }
+                c = name->val.p[i];
+                lowbits = (c & 0x0F);
+                highbits = c >> 4;
+                s[j++] = nibble_to_hex_digit(highbits);
+                s[j++] = nibble_to_hex_digit(lowbits);
+            }
+#else
+            return MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE;
```
[the remark this answers is not in the document]
Comment (AgathiyanB) [reply library/x509.c:911]: Would you think it would be better to change this line to `MBEDTLS_ERR_X509_INVALID_NAME` or to just make the whole function not work at all if ASN1_WRITE_C is disabled?

[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply library/x509.c:911]: Hmm, good question. What would happen with such an input previously? And I note that this is a "get" function - it seems weird that its behaviour depends upon a "write" configuration parameter. Maybe the correct fix here is to allow the needed functions even if we don't have ASN1_WRITE_C?

@davidhorstmann-arm you've spent more time in the ASN.1 modules than I have - do you have an opinion here?

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/x509.c:911]: > allow the needed functions even if we don't have ASN1_WRITE_C?

I don't understand what this code is doing, but yes, this seems like the right thing.

Why is parsing code calling write functions in the first place? This code needs comments explaining what's going on.

[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply library/x509.c:911]: > do you have an opinion here?

I think we should leave it as-is. Let me give a quick background of what is going on here:

In cases where a Relative Distinguished Name is not a string, RFC 4514 allows us to output the RDN as a string in the form:
```
<OID number>=#<hex encoding of raw der>
```
For example:
```
1.3.6.1.4.1.1466.0=#04024869
```
We need `MBEDTLS_ASN1_WRITE_C` to encode the DER data that gets put after the `#`, because we don't store it in its raw form but unpacked in a struct.

Most RDNs are strings and don't need this alternate form, so ideally we should allow users to use this function without needing `MBEDTLS_ASN1_WRITE_C`. When we encounter a name that is not a string, we return `MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE`, as in "the feature of printing non-string RDNs is unavailable, because you didn't enable `MBEDTLS_ASN1_WRITE_C`".

However, I'm not sure whether this aligns with the way we use `FEATURE_UNAVAILABLE` normally, as this is not a clearly delineated "X509 Feature". Do you think that `INVALID_NAME` would be a better choice?

In any case, it looks like we at least need some code comments here, explaining what is going on and why we need `MBEDTLS_ASN1_WRITE_C`. Additionally, we could do with a note in the doxygen comment in the header, something along the lines of:
```doxygen
\note           Converting names with arbitrary non-string RDNs
                requires MBEDTLS_ASN1_WRITE_C to be enabled.
```

[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply library/x509.c:911]: Thanks, that matches with what my understanding was.

However, if we're going to enhance the parsing of ASN.1 data, having it depend on a "WRITE" configuration value just seems wrong.

> Most RDNs are strings and don't need this alternate form, so ideally we should allow users to use this function without needing `MBEDTLS_ASN1_WRITE_C`.

The issue is that the type of RDN that the user is going to pass to this function isn't necessarily known at compile time, so having it depend on a compile-time value is sub-optimal.

> we don't store it in its raw form but unpacked in a struct.

That's our choice/problem, so shouldn't make the user's life harder.

> we could do with a note in the doxygen comment in the header, something along the lines of ...

That would have made review easier :) but I don't think it's the right approach here.

[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply library/x509.c:911]: So I think those functions needed by this function need to be enabled by `MBEDTLS_X509_USE_C`as well as `MBEDTLS_ASN1_WRITE_C`

[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply library/x509.c:911]: > The issue is that the type of RDN that the user is going to pass to this function isn't necessarily known at compile time, so having it depend on a compile-time value is sub-optimal.

I think the easiest way to do this would be to duplicate the functionality of the `asn1_write` functions locally in this file. They are quite simple, so this is definitely doable without too much risk. Would this be a workable solution?

[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply library/x509.c:911]: Sorry, I disagree: `mbedtls_asn1_write_len()` alone is 63 lines of C, and 186 bytes of code in Thumb2 - that's not something we should be duplicating at all

[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply library/x509.c:911]: Okay that's fair enough, I suppose they are larger than I remembered.

Ah I see now what you mean - apologies, I completely misunderstood. You propose that we should move the definitions of `mbedtls_asn1_write_tag()` and `mbedtls_asn1_write_len()` so that they are included even without `MBEDTLS_ASN1_WRITE_C`?

This solution seems fine to me, and I agree that it's the best thing we can do in the circumstances.

[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply library/x509.c:911]: > You propose that we should move the definitions [...] so that they are included even without MBEDTLS_ASN1_WRITE_C (but with MBEDTLS_X509_USE_C).

Yes - they now become part of the "core" X.509 code, rather than functions only needed for WRITE.

Some careful testing will be necessary to ensure that the library still builds under various combination of config options

**Comment by AgathiyanB** [comment]
> In addition to those comments, I think it would be good to have some "round-trip" tests, especially with characters that get escaped

`mbedtls_x509_string_to_names` works as a round trip test as it takes strings using string to names and then get backs out the output using dn_gets. I will add tests to serve more to test round trip functionality rather than just testing the functionality of `mbedtls_x509_string_to_names` but just confirming these are the only round trip tests you wanted added.

**Comment by tom-cosgrove-arm** [comment]
> just confirming these are the only round trip tests you wanted added

That's fine, thanks

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
2 small things, looks good otherwise

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, thanks!

File: library/x509.c [hunk 15]
```diff
@@ -840,32 +851,91 @@ int mbedtls_x509_dn_gets(char *buf, size_t size, const mbedtls_x509_name *dn)
...
-            if (j >= sizeof(s) - 1) {
-                return MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
-            }
+        if (print_hexstring) {
+            s[0] = '#';
 
-            c = name->val.p[i];
-            // Special characters requiring escaping, RFC 1779
-            if (c && strchr(",=+<>#;\"\\", c)) {
+            asn1_len_p = asn1_tag_len_buf + 10;
```
Comment (tom-cosgrove-arm) [inline library/x509.c:877]: Where does this "magic value" 10 come from?

Comment (tom-cosgrove-arm) [reply library/x509.c:877]: I mean, it should probably be `sizeof(asn1_tag_len_buf)`, but why is `10` big enough?

Comment (AgathiyanB) [reply library/x509.c:877]: From the size of the buffer in the declaration of it (although I will add a variable for this size to replace the instances of 10).

Comment (tom-cosgrove-arm) [reply library/x509.c:877]: If it's just `sizeof(asn1_tag_len_buf)`, then use that expression rather than a new variable, but have a comment where `asn1_tag_len_buf` is declared explaining why it's that size

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
> changelog in a future PR

@davidhorstmann-arm will you prepare a PR with a ChangeLog entry?
